### DISCUSSION TITLE
RangeOfT enhancements for simplicity to consumers

### DISCUSSION THREAD

**Discussion by dcuccia** [VirtualPhotonics/VTS#81] [Ideas]
I've found, in creating the user samples, the utility of having extension method operators on `Range<T>` objects. The need to call `myRange.AsEnumerable().ToArray()` is unconcise and not very intuitive to newcomers. It will also be that much more confusing if we need to add any additional calls to make it compatible with PythonNet. 

In order to resolve these, I'd propose:

1. at a minimum, add a `ToArray()` extension method in a new `RangeExtensions` class
2. consider instead making the range itself implement IEnumerable, so that someone can benefit from all the LINQ goodness directly. E.g., they could say: 
```csharp
double[] endpoints = myRange.ToArray();
List<double> endpoints = myRange.ToList();
foreach(double endpoint in myRange) { //... }
```

This would also clean up the code significantly throughout the codebase. Only check would be to ensure that any WPF databinding to a Range isn't disrupted by it newly implenting `IEnumerable<T>` (sometimes there are weird defaults, but this one is easy to check that databinding continues to work).

**Comment by hayakawa16** [comment]
Hi, I just pulled master into the branches because so much has been merged to master lately.  I'm on branch ch-210312-raytrace-MC and when I build I get 2 warnings:
![Screenshot 2023-08-31 155402](https://github.com/VirtualPhotonics/VTS/assets/26939347/0616183f-088a-43d6-9644-ed7254043d66)

Did I merge incorrectly?

**Reply by lmalenfant** [reply]
Hi Carole, 

You merged correctly, one of the warnings is because David marked AsEnumerable as obsolete to encourage users to use the built in IEnumerable. We will remove this eventually.

The other is in one of the scripts, @dcuccia could you take a look and see if we fix that one?

**Reply by dcuccia** [reply]
Will do!

**Comment by lmalenfant** [comment]
This has been implemented.

**Reply by hayakawa16** [reply]
I just pulled latest master and still see the Warning above when I build.

**Reply by dcuccia** [reply]
Sorry, thought I responded to this. This is by design. If you update the code that is generating this warning (remove the call to the obsolete extension method), it will go away.
